### PR TITLE
spanconfigccl: fix tests under multitenancy

### LINKED ISSUE #106818: spanconfigsplitterccl: TestDataDriven is brittle (perhaps due to weakness in spanconfigtestcluster? unsure)
Informs #76378 

**Describe the problem**

When `TestDataDriven` is run with a random test tenant server assigned, the test crashes with a nil pointer exception.

It's unclear what is happening. 

**To Reproduce**

Replace the use of `TestDoesNotWorkWithSecondaryTenantsButWeDontKnowWhyYet` by `TestTenantAlwaysEnabled` and run the test.

**Expected behavior**

From a **test quality** perspective, if the test code lands in an unsupported configuration it should fail with `t.Fatal` not with a crash.

Jira issue: CRDB-29728

### LINKED ISSUE #106821: spanconfigsqlwatcherccl: TestSQLWatcherMultiple gets confused watching over secondary tenants
Informs #76378

**Describe the problem**

When `TestSQLWatcherMultiple` is run over a secondary tenant, it fails some of its conditions, e.g.:
```
    sqlwatcher_test.go:345:
                Error Trace:    pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/spanconfigsqlwatcherccl_test/pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go:345
                                                        GOROOT/src/runtime/asm_amd64.s:1594
                Error:          Not equal:
                                expected: 1
                                actual  : 0
                Test:           TestSQLWatcherMultiple
```

**How to reproduce**

Replace `TestDoesNotWorkWithSecondaryTenantsButWeDontKnowWhyYet` by `TestTenantAlwaysEnabled` and run the test.

**Expected solution**

The test code should function and succeeds whether or not the watcher is run over a secondary tenant.

Jira issue: CRDB-29730

### PR DESCRIPTION
**Pull request by rafiss**
fixes https://github.com/cockroachdb/cockroach/issues/106818
fixes https://github.com/cockroachdb/cockroach/issues/106821
Release note: None

### CODE REVIEW

**Review by yuzefovich — APPROVED** [review]
:lgtm: Thanks!

Also, take a look at `TestTODOTestTenantDisabled` in case you can remove some files from the allowlist.

> Reviewed 2 of 2 files at r1, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107760#-:-NaOQbQe79mws_ZPsFbh:b-q5b6zn)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

___
*[`pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go` line 290 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107760#-NaOQ8n1E9eSYeaZ7nuk:-NaOQ8n1E9eSYeaZ7nul:b-on2a0p) ([raw file](https://github.com/cockroachdb/cockroach/blob/44c8346a2aee385de9b364a2d3dd444d35b3b5d9/pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go#L290)):*
> ```Go
> 	tc := testcluster.StartTestCluster(t, 1, base.TestClusterArgs{
> 		ServerArgs: base.TestServerArgs{
> 			DefaultTestTenant: base.TestTenantProbabilisticOnly,
> ```

nit: you can omit specifying `DefaultTestTenant` since "probabilistic only" is the default.
___
*[`pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go` line 308 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107760#-NaOQLjyCxUfm8dfsJpa:-NaOQLjyCxUfm8dfsJpb:b-87n0h0) ([raw file](https://github.com/cockroachdb/cockroach/blob/44c8346a2aee385de9b364a2d3dd444d35b3b5d9/pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go#L308)):*
> ```Go
> 	noopCheckpointDuration := 100 * time.Millisecond
> 	sqlWatcher := spanconfigsqlwatcher.New(
> 		ts.TenantOrServer().Codec(),
> ```

nit: FYI `TenantOrServer()` is being renamed to `ApplicationLayer()` over in #107612.

**Review by yuzefovich — COMMENTED** [review]
*at `TestTODOTestTenantDisabled` linter.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107760#-:-NaOR7DT9kYoVu0RZLQn:bnuorna)* status: :shipit: complete! 1 of 0 LGTMs obtained (waiting on @rafiss)

**Review by rafiss — COMMENTED** [review]
hm, nothing in spanconfigccl is excluded from that linter at all, so i'm not sure why the linter was passing before. tftr!

bors r=yuzefovich

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/107760#-:-NaP-QBA-IFNgqq6G8-B:b-j9880w)* status: :shipit: complete! 0 of 0 LGTMs obtained (and 1 stale) (waiting on @yuzefovich)

___
*[`pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go` line 290 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/107760#-NaOQ8n1E9eSYeaZ7nuk:-NaP-PUCDNB5lBJPUZPb:bb7f2mf) ([raw file](https://github.com/cockroachdb/cockroach/blob/44c8346a2aee385de9b364a2d3dd444d35b3b5d9/pkg/ccl/spanconfigccl/spanconfigsqlwatcherccl/sqlwatcher_test.go#L290)):*
<details><summary><i>Previously, yuzefovich (Yahor Yuzefovich) wrote…</i></summary><blockquote>

nit: you can omit specifying `DefaultTestTenant` since "probabilistic only" is the default.
</blockquote></details>

done

**Comment by rafiss** [comment]
ah i see https://github.com/cockroachdb/cockroach/pull/107612 is merging right now

bors r-

File: pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go [hunk 1]
```diff
@@ -119,7 +113,8 @@ func TestDataDriven(t *testing.T) {
 					var dbName, tbName string
 					d.ScanArgs(t, "database", &dbName)
 					d.ScanArgs(t, "table", &tbName)
-					objID = tenant.LookupTableByName(ctx, dbName, tbName).GetID()
+					tbl := tenant.LookupTableByName(ctx, dbName, tbName)
+					objID = tbl.GetID()
```
Comment (knz) [inline pkg/ccl/spanconfigccl/spanconfigsplitterccl/datadriven_test.go:117]: nit: Jake introduced a method .QueryTableID() which might be relevant here.

**Comment by rafiss** [comment]
bors r=yuzefovich

**Comment by yuzefovich** [comment]
> hm, nothing in spanconfigccl is excluded from that linter at all, so i'm not sure why the linter was passing before.

The linter is specific to `base.TODOTestTenantDisabled`, and in this case we had `base.TestDoesNotWorkWithSecondaryTenantsButWeDontKnowWhyYet`. So all good.

**Pull request merged**